### PR TITLE
Simplify Bmp Decoding / Encoding

### PR DESCRIPTION
**Pull request by iamcarbon**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
- Eliminates the EndianBinaryWriter from the BMP encoder. 
- Writes the packed BmpHeader & BmpInfoHeader directly to the buffer

### CODE REVIEW

**Comment by iamcarbon** [comment]
OK, this is ready for review.

Working to kill the EndianBinary(Reader / Writer) now that we have all these nice new memory helpers.

File: src/ImageSharp/Formats/Bmp/BmpInfoHeader.cs [hunk 1]
```diff
@@ -8,28 +12,55 @@ namespace SixLabors.ImageSharp.Formats.Bmp
     /// the screen.
     /// <see href="https://en.wikipedia.org/wiki/BMP_file_format"/>
     /// </summary>
-    internal sealed class BmpInfoHeader
+    [StructLayout(LayoutKind.Sequential, Pack = 1)]
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Bmp/BmpInfoHeader.cs]: Clever!

File: src/ImageSharp/Formats/Bmp/BmpEncoderCore.cs [hunk 2]
```diff
@@ -54,96 +50,58 @@ public void Encode<TPixel>(Image<TPixel> image, Stream stream)
...
             var fileHeader = new BmpFileHeader(
                 type: 19778, // BM
                 offset: 54,
                 reserved: 0,
                 fileSize: 54 + infoHeader.ImageSize);
 
-            WriteHeader(writer, fileHeader);
-            this.WriteInfo(writer, infoHeader);
-            this.WriteImage(writer, image.Frames.RootFrame);
+            byte[] buffer = new byte[40]; // TODO: stackalloc
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Bmp/BmpEncoderCore.cs]: It's probably better to actually do the stackalloc now. We'll forget otherwise.

Comment (iamcarbon) [reply src/ImageSharp/Formats/Bmp/BmpEncoderCore.cs]: Annoyingly, we can't use stack allocated memory (prior to NETCORE 2.1) for the stream reads. Removing these comments for now.

File: src/ImageSharp/Formats/Bmp/BmpFileHeader.cs [hunk 3]
```diff
@@ -44,12 +48,25 @@ public BmpFileHeader(short type, int fileSize, int reserved, int offset)
...
         /// </summary>
         public int Offset { get; }
+
+        public static BmpFileHeader Parse(Span<byte> data)
+        {
+            return MemoryMarshal.Cast<byte, BmpFileHeader>(data)[0];
+        }
+
+        public unsafe void WriteTo(Span<byte> buffer)
+        {
```
Comment (antonfirsov) [inline src/ImageSharp/Formats/Bmp/BmpFileHeader.cs]: How about this without pinning? 
```C#
ref BmpFileHeader dest = ref Unsafe.As<byte, BmpFileHeader>(ref MemoryMarshal.GetReference(buffer));
dest = this;
```
Doesn't really matter here, but it's most likely faster :)

File: src/ImageSharp/Formats/Bmp/BmpInfoHeader.cs [hunk 4]
```diff
@@ -91,5 +122,40 @@ internal sealed class BmpInfoHeader
...
+
+        /// <summary>
+        /// Parses the BITMAPCOREHEADER consisting of the headerSize, width, height, planes, and bitsPerPixel fields (12 bytes).
+        /// </summary>
+        /// <param name="data">The data to parse,</param>
+        /// <returns>Parsed header</returns>
+        /// <seealso href="https://msdn.microsoft.com/en-us/library/windows/desktop/dd183372.aspx"/>
+        public static BmpInfoHeader ParseCore(ReadOnlySpan<byte> data)
+        {
+            return new BmpInfoHeader(
```
Comment (antonfirsov) [inline src/ImageSharp/Formats/Bmp/BmpInfoHeader.cs]: Just wondering why don't you do the casting trick here?

Comment (iamcarbon) [reply src/ImageSharp/Formats/Bmp/BmpInfoHeader.cs]: The core header is only 12 bytes -- so  the cast won't work here. Adding one more check for the full cast case to make sure we're always passing 40 bytes...

File: src/ImageSharp/Formats/Bmp/BmpFileHeader.cs (hunk 3 above)
Comment (iamcarbon) [reply src/ImageSharp/Formats/Bmp/BmpFileHeader.cs]: Much nicer!

**Comment by iamcarbon** [comment]
@antonfirsov Good feedback w/ suggestions applied!

**Comment by antonfirsov** [comment]
Well done! Should we merge this when appveyor is finished?

**Comment by iamcarbon** [comment]
Travis, are you ok?

**Pull request merged**